API specification

kairoaraujo opened the pull request:
This commit implements the API specification.
The specification can be read easily on
http://kairo.eti.br/ownca-rest-api/ or running the local HTTP
API locally, after merge it will be available on
http://ownca.github.io/ownca-rest-api/ as well.
```
pipenv shell
pipenv install
flask run
```

This commit has no functional implementation, and the idea is to
review the API specification until it gets ready and then create
the issues related to the implementation.

Signed-off-by: Kairo de Araujo <kairo@dearaujo.nl>

kairoaraujo committed:
API specification

This commit implements the API specification.
The specification can be read easily on
http://kairo.eti.br/ownca-rest-api/ or running the local HTTP
API locally, after merge it will be available on
http://github.io/ownca-rest-api/ as well.
```
pipenv shell
pipenv install
flask run
```

This commit has no functional implementation, and the idea is to
review the API specification until it gets ready and then create
the issues related to the implementation.

Signed-off-by: Kairo de Araujo <kairo@dearaujo.nl>

lawndoc commented on docs/swagger.json:
```diff
@@ -0,0 +1,549 @@
+                    "in": "path",
+                    "required": true,
+                    "type": "string"
+                }
```
Shouldn't we also add optional parameters to specify other parts of the certificate (oids, dns_names, etc...)?

lawndoc replied:
Looking at the [rendered documentation](http://kairo.eti.br/ownca-rest-api/#/%2Fca/get_certificate_authority_management) it does show all the fields in the example response, but it should also show them in the parameters if they can be specified in the request.

kairoaraujo replied:
> Shouldn't we also add optional parameters to specify other parts of the certificate (oids, dns_names, etc...)?

Required by default is False. Do you mean ben explicitly with the ‘’required=False’’?

kairoaraujo replied:
> Looking at the [rendered documentation](http://kairo.eti.br/ownca-rest-api/#/%2Fca/get_certificate_authority_management) it does show all the fields in the example response, but it should also show them in the parameters if they can be specified in the request.

I don't follow. Can you explain? The request is made as a payload (JSON) in the body. Sorry, maybe I missed some detail. :)

lawndoc replied:
Sorry I probably just didn't give enough detail. In the rendered documentation, I'm looking at POST request to `/api/v1/ca/{ca_common_name}/certificates`, and it shows `ca_common_name` as the only parameter for this endpoint.

It doesn't mention how to send fields like `oids` or `dns_names` in the request when creating a new certificate. In the "Responses" section, it shows all of the certificate fields, but don't we also want to be able to use them in the request?

For example, if I wanted to issue a new certificate with the `dns_names` set to `["mysite.example.com", "myothersite.example.com"]`, I would need to have a parameter for `dns_names` when I send a request to `/api/v1/ca/{ca_common_name}/certificates`

kairoaraujo replied:
That was a good catch! Thanks!
I also fixed the response body.

What do you think about the ``required=False``, do we need to add it explicitly 🤔

lawndoc replied:
I think it doesn't affect readability much either way, so we could just add it explicitly in case someone wants it for some reason while ingesting the spec.

lawndoc replied:
Well, I don't think it would be a big deal either way. Your call

kairoaraujo replied:
That is a change that, if needed, we can work on further.

lawndoc reviewed (changes requested):
Found one thing missing in the schema. The endpoint to issue a new certificate is missing the optional parameters.

lawndoc commented:
You're a coding machine haha! I really enjoy seeing how you're constructing the flask API with flask-restx and the docs with swagger. I've never ran across those before, but it's a very cohesive setup for developers. Thanks for the learning opportunity :)

kairoaraujo commented:
> You're a coding machine haha! I really enjoy seeing how you're constructing the flask API with flask-restx and the docs with swagger. I've never ran across those before, but it's a very cohesive setup for developers. Thanks for the learning opportunity :)

Hahaha. I would like to have more time. 😀 
That is the idea here, have fun, learn and leave something to help others. 😁 
Thank you. 

The idea to have a defined specification for the API will prevent us from spending time in code. Having an API first will speed up the implementation. I'm glad you liked it.

kairoaraujo committed:
Fix the endpoint for issuing certificate

It was missing the body documentation as the response body wasn't in the
good format ``{data: {}, 'error': {}}``

Signed-off-by: Kairo de Araujo <kairo@kairo.eti.br>

kairoaraujo commented:
Thanks @lawndoc for reviewing.
